### PR TITLE
adds ability for users to make and reply to comments

### PR DESCRIPTION
**Pull request by hernanat**
https://trello.com/c/Jj1W2US4
https://trello.com/c/LetQxTrG

This commit allows users to make comments / reply to them
  as well as format their comments using markdown and LaTex.

A subsequent commit will cleanup the associated queries, and
  come up with an efficient way for preparing comment trees
  for display.

### CODE REVIEW

**Comment by hernanat** [comment]
@ashpil I'm struggling a bit with the comment formatting here. Any ideas?

Screen Shot 2020-05-18 at 1 13 02 PM

**Comment by ashpil** [comment]
Right, that's because the first line if offset by the vote controls, but next lines aren't so they start from where the div starts. I'll push a commit with some smoother formatting.

**Comment by ashpil** [comment]
What's the purpose of the `.comment-tree-line` class?

**Comment by hernanat** [comment]
some of it probably has to do with the markdown too huh. I know we have some overriding styles in our application default

**Comment by hernanat** [comment]
> What's the purpose of the .comment-tree-line class?

That's the left line bar you see. If everything was formatted correctly it would run along the left hand side of the comment body. This will be helpful when we have comment subtrees being displayed

**Comment by ashpil** [comment]
I don't think you need a separate class for that, I imagine it could be done with just a CSS border?

**Comment by hernanat** [comment]
> I don't think you need a separate class for that, I imagine it could be done with just a CSS border?

Maybe. I tried that at first but it went all the way to the left past the hide / vote controls

**Comment by hernanat** [comment]
Screen Shot 2020-05-18 at 2 35 58 PM

I also just added reply boxes :)

**Comment by hernanat** [comment]
Screen Shot 2020-05-18 at 3 03 37 PM

**Comment by ashpil** [comment]
Here's what this commit does:

image

I styled the comments using grid, I think this is cleaner though I'm not yet sure if this should be the final style.

Markdown wasn't working in comments just because they didn't have the markdown class. I created a "markdownable" class and put markdown in a separate SCSS file.

For some reason, replying to comments doesn't seem to be working for me. I might be doing it wrong.

**Comment by hernanat** [comment]
dude this is sick, thanks! I made a few changes, here's what they look like:

Screen Shot 2020-05-18 at 3 45 29 PM

I'll push the updates now so you can play around locally with replies.

Also -- I haven't optimized the comment querying yet as I just want to get the actual actions out of the way, I'll take care of better querying in a separate commit.

File: app/assets/stylesheets/pages/_submissions.scss [hunk 1]
```diff
@@ -157,3 +94,76 @@
...
+      }
+      a {
+        text-decoration: none;
+      }
+    }
+    .comment-body {
+      grid-column: 3 / 3;
+      grid-row: 2 / span 2;
+      display: block;
+      :first-child {
```
Comment (hernanat) [inline app/assets/stylesheets/pages/_submissions.scss:142]: there was a weird gap between the byline and the body due to the margins defined in the markdown styling

Comment (ashpil) [reply app/assets/stylesheets/pages/_submissions.scss:142]: Actually, I think this should be reflected in the markdown for comments as well. I'll push a fix.

**Comment by hernanat** [comment]
Screen Shot 2020-05-18 at 6 06 55 PM

is there an easy way to not spread them out so much? Or to make them closer when the screen size shrinks?  It appears that if I change the grid settings it either flings them apart or squeezes them completely together :(

**Comment by ashpil** [comment]
> is there an easy way to not spread them out so much?

Them being?

**Comment by hernanat** [comment]
> Them being?

sorry -- the nested comments. I guess I mean to make them not so indented?

**Comment by hernanat** [comment]
and is there a way to get them to wrap on small screens? i.e. this is what it looks like now:
Screen Shot 2020-05-18 at 6 15 01 PM

also if you don't have time right now feel free to point me in a direction and I can try to figure it out :)

**Comment by ashpil** [comment]
Nah, I can take a look at it. There's some unneeded padding in the subcomments, and I'm trying to figure out the best way to squeeze even more room in.

**Comment by hernanat** [comment]
ahh ok cool thanks. I'll start writing the tests for this :)

**Comment by ashpil** [comment]
Should be better now.

image

**Comment by hernanat** [comment]
sweet I think that this is all set @ashpil let me know what you think

**Comment by hernanat** [comment]
[As a user I can comment on a submission](https://trello.com/c/Jj1W2US4/3-as-a-user-i-can-comment-on-a-submission)

**Comment by hernanat** [comment]
[As a user I can format a submission / comment using markdown](https://trello.com/c/LetQxTrG/8-as-a-user-i-can-format-a-submission-comment-using-markdown)

**Comment by ashpil** [comment]
What are your thoughts on this?

image

Switching the hide and votes buttons (I think this is how Reddit does it). It's certainly much more condensed, though I'm not sure if this is too dense and thus unapproachable.

**Comment by hernanat** [comment]
I'd be down to try that out, if our users don't end up liking it we can revisit it then ~

**Comment by ashpil** [comment]
For some reason it was giving me a nasty merge error, but I think I fixed it all. Don't pull before you make sure.

**Comment by hernanat** [comment]
Git strikes again 😂 sorry I had squashed everything down before

**Comment by hernanat** [comment]
You actually re-pushed the commits I squished down hehe

**Comment by ashpil** [comment]
Whoops lol sorry

**Comment by hernanat** [comment]
All good! Ill fix it up when Im at my computer. Thanks again!

**Comment by hernanat** [comment]
@ashpil should be fixed up and ready to go

**Comment by hernanat** [comment]
thanks @ashpil ! and thanks for taking the reigns on the design stuff!

**Comment by ashpil** [comment]
[As a user I can format a submission / comment using LaTeX](https://trello.com/c/EBpIRrJ8/9-as-a-user-i-can-format-a-submission-comment-using-latex)

**Pull request merged**